src/main/java/org/broadinstitute/hellbender/tools/ConvertHadoopBamFragmentToBam.java

```diff
@@ -0,0 +1,88 @@
+import org.broadinstitute.hellbender.cmdline.CommandLineProgramProperties;
+import org.broadinstitute.hellbender.cmdline.programgroups.TestSparkProgramGroup;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.io.IOUtils;
+
+import java.io.*;
+
+
+/**
+ * Spark troubleshooting utility that converts a Hadoop bam fragment (eg., a part0000, part0001, etc. file
+ * produced by ReadsSparkSink) into a readable bam file by adding a header and a BGZF terminator.
+ */
+@CommandLineProgramProperties(
+        summary = "Convert a hadoop bam fragment into a readable bam",
+        oneLineSummary = "Convert a hadoop bam fragment into a readable bam",
+        programGroup = TestSparkProgramGroup.class
+)
+public final class ConvertHadoopBamFragmentToBam extends CommandLineProgram {
+
+    @Argument(shortName = "hadoopFragment", fullName = "hadoopFragment", doc = "Hadoop BAM fragment to be converted", optional = false)
+    private File hadoopFragment;
+
+    @Argument(shortName = "bamWithHeader", fullName = "bamWithHeader", doc = "Well-formed BAM whose header to use for the converted fragment", optional = false)
+    private File bamWithHeader;
+
+    @Argument(shortName = "O", fullName = "outputBam", doc = "Location to write the converted BAM Fragment", optional = false)
+    private File outputBam;
+
+    @Override
+    protected Object doWork(){
+        SAMFileHeader header = null;
+        try ( final SamReader headerReader = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT).open(bamWithHeader) ) {
+            header = headerReader.getFileHeader();
+        }
+        catch ( IOException e ) {
+            throw new UserException("Error reading header from " + bamWithHeader.getAbsolutePath(), e);
+        }
+
+        try ( FileOutputStream outStream = new FileOutputStream(outputBam) ) {
+            writeHeader(outStream, header);
+            final byte[] fragmentBytes = IOUtils.readFileIntoByteArray(hadoopFragment);
+            outStream.write(fragmentBytes);
+            outStream.write(BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK);
+        }
+        catch ( IOException e ) {
+            throw new UserException("Error writing to " + outputBam.getAbsolutePath(), e);
+        }
+
+        return null;
+    }
+
+    private static void writeHeader(final OutputStream outputStream, final SAMFileHeader samFileHeader) {
+        final BlockCompressedOutputStream blockCompressedOutputStream = new BlockCompressedOutputStream(outputStream, null);
+        final BinaryCodec outputBinaryCodec = new BinaryCodec(new DataOutputStream(blockCompressedOutputStream));
+
+        final String headerString;
+        final Writer stringWriter = new StringWriter();
+        new SAMTextHeaderCodec().encode(stringWriter, samFileHeader, true);
+        headerString = stringWriter.toString();
+
+        outputBinaryCodec.writeBytes("BAM\1".getBytes());
```

Comment:
Yes. It's defined in htsjdk, but the definition is package protected, so we have to duplicate it here.